src/InstrumentationEngine/ProfilerManager.h

```diff
@@ -368,15 +368,9 @@ namespace MicrosoftInstrumentationEngine
 
             CComPtr<TInterfaceType> pCallback;
 
-            // Holding the lock during the callback functions is dangerous since rentrant
-            // events and calls will block. Copy the collection under the lock, then release it and finally call the callbacks
+            if (m_profilerCallbackHolder != nullptr)
             {
-                CCriticalSectionHolder lock(&m_cs);
-
-                if (m_profilerCallbackHolder != nullptr)
-                {
-                    pCallback = (TInterfaceType*)(m_profilerCallbackHolder->GetMemberForInterface(__uuidof(TInterfaceType)));
-                }
+                pCallback = (TInterfaceType*)(m_profilerCallbackHolder->GetMemberForInterface(__uuidof(TInterfaceType)));
```

Comment:
I think this is safe since m_profilerCallbackHolder should just get initialized in the Initialize method and either become null or stay non-null for the rest of its lifetime. If that doesn't end up being the case we could also use Interlocked functions instead of critical section.

Reply:
Per our discussion, if we want to maintain the RemoveRawProfilerHook, we do need to synchronize m_profilerCallbackHolder, just not with a critical section.

Reply:
We discussed several options offline:

atomic shared_ptr provides thread safety only when writing against different shared_ptr instances. Writing to the same shared_ptr instance must be exclusively locked:
```
p1 = p; p2 = p; // allowed, Reads of p
p3 = p1; p1 = nullptr; // not allowed, Read/Write of p1
p2 = p; p2 = nullptr; // not allowed, multiple Writes of p2
```

atomic bool guards are not sufficient as atomic bools do not guarantee the pointer safety:
```
// Thread 1
if (atomic_bool) { /* use ptr */ }
// Thread 2
atomic_bool = false
/* modify ptr */

// we may succeed the check in Thread 1 but before using ptr, Thread 2 might set the atomic bool and modify ptr
```

Reply:
I've replaced critical_sections with unique_lock (for writes) and shared_lock (for reads) on a shared_mutex.

Reply:
Have you tried building these on Linux? I'm not certain the C++ library that our version of clang understands is new enough for these.

Reply:
I think this is a good solution, but we'll have to validate the performance.

Reply:
atomic shared_ptr functions compile so I'll go ahead with that implementation.

Reply:
FYI, these will be replaced with atomic<shared_ptr> in C++ 20, not that we will be switching to that any time soon.

---
In reply to: [546068525](https://github.com/microsoft/CLRInstrumentationEngine/pull/338#discussion_r546068525) [](ancestors = 546068525)